torch_xla/distributed/zero_redundancy_optimizer.py

```diff
@@ -256,26 +270,52 @@ def step(self, closure=None, **kwargs):
 
     # Reduce full gradients across ranks
     # Assign gradient shards to the respective parameter shards
+    padded_grads = []
     for param_group, sharded_param_group in zip(
         self.param_groups, self.base_optimizer.param_groups):
       for param, shard in zip(param_group['params'],
                               sharded_param_group['params']):
         if param.grad is not None:
           padded_grad = self._pad_to_world_size(param.grad,
                                                 self.local_world_size)
-          grad_shard = xm.reduce_scatter(
-              xm.REDUCE_SUM,
-              padded_grad,
-              scale=1.0 / self.local_world_size,
-              scatter_dim=0,
-              shard_count=self.local_world_size,
-              pin_layout=self.pin_layout,
-              groups=self.sharding_groups,
-          )
-
-          if grad_shard.dtype != self.optimizer_dtype:
-            grad_shard = grad_shard.to(dtype=self.optimizer_dtype)
+          if self.coalesce_cc:
+            padded_grads.append(padded_grad)
+          else:
+            grad_shard = xm.reduce_scatter(
+                xm.REDUCE_SUM,
+                padded_grad,
+                scale=1.0 / self.local_world_size,
+                scatter_dim=0,
+                shard_count=self.local_world_size,
+                pin_layout=self.pin_layout,
+                groups=self.sharding_groups,
+            )
+            if grad_shard.dtype != self.optimizer_dtype:
+              grad_shard = grad_shard.to(dtype=self.optimizer_dtype)
+            shard.grad = grad_shard
+
+    if self.coalesce_cc:
+      grad_shards = xm.reduce_scatter(
+          xm.REDUCE_SUM,
+          padded_grads,
+          scale=1.0 / self.local_world_size,
+          scatter_dim=0,
+          shard_count=self.local_world_size,
+          pin_layout=self.pin_layout,
+          groups=self.sharding_groups,
+      )
+      index = 0
+      for param_group, sharded_param_group in zip(
+          self.param_groups, self.base_optimizer.param_groups):
+        for param, shard in zip(param_group['params'],
+                                sharded_param_group['params']):
+          if param.grad is not None:
+            grad_shard = grad_shards[index]
+
+            if grad_shard.dtype != self.optimizer_dtype:
+              grad_shard = grad_shard.to(dtype=self.optimizer_dtype)
           shard.grad = grad_shard
```

Comment:
this 2 line not in `if param.grad is not None:`, need indents

Reply:
fixed. thanks.